### ISSUE TITLE
exabgp busyspin after some time

### ISSUE THREAD

**Issue by chantra** [Exa-Networks/exabgp#723]
##### ISSUE TYPE
 - Bug Report

##### OS
CentOS 7
Linux 4.6

##### VERSION

```
ExaBGP : 4.0.2-1c737d99
Python : 2.7.11 (default, May 29 2017, 05:07:11)  [GCC 5.x 20170328]
Uname  : Linux hostname 4.6.7-xxxxxx #53 SMP Sun Jul 9 06:32:52 PDT 2017 x86_64
Root   :
```

##### ENVIRONMENT 
```
exabgp.daemon.pid=/path/to/file.pid
```

##### CONFIGURATION
```
template {
    neighbor all {
        router-id 1.2.3.4;
        local-as 1234;
        hold-time 20;
        capability {
            graceful-restart;
        }
    }
}

process myprocess {
  run /path/to/script;
  encoder text;
}

neighbor 1.2.3.5 {
    inherit all;
    local-address 1.2.3.4;
    family {
        ipv4 unicast;
    }
    peer-as 1235;
}

```

##### SUMMARY

After some time exabgp process will run 100% CPU.

##### STEPS TO REPRODUCE

I do not have a good way to reproduce yet but filing this for awareness.

##### EXPECTED RESULTS

Exabgp should not busy spin.

##### Remarks:

On a system not busy spinning (yet), here is what a strace looks likes:

```
wait4(168, 0x7ffc54ffa43c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
select(11, [4 7 9 10], [], [], {0, 10000}) = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(167, 0x7ffc54ffa43c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
wait4(168, 0x7ffc54ffa43c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
select(11, [4 7 9 10], [], [], {0, 10000}) = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(167, 0x7ffc54ffa43c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
wait4(168, 0x7ffc54ffa43c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
select(11, [4 7 9 10], [], [], {0, 10000}) = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
```

**mind** the select waiting on all file descriptor with a timeout of 10ms.

On a system busy spinning:

```
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(157, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
wait4(156, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(157, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
wait4(156, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(157, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
wait4(156, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(157, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
wait4(156, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
wait4(157, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(10, [9], [], [], {0, 0})         = 0 (Timeout)
wait4(156, 0x7fff80d7945c, WNOHANG, NULL) = 0
select(8, [7], [], [], {0, 0})          = 0 (Timeout)
select(5, [4], [], [], {0, 0})          = 0 (Timeout)
select(11, [10], [], [], {0, 0})        = 0 (Timeout)
```

The pattern is similar, but we never get the:
```
select(11, [4 7 9 10], [], [], {0, 10000}) = 0 (Timeout)
```

part.

Mind that this is different from #704 and 24fca1cbc91f1d768478217cb9efd9af40b01561 is not applied (which may just make it worth) 
The issue seems to be due to `self.async.ready()` always returning True.

**Comment by thomas-mangin** [comment]
If `self.async.ready()` returns true to would mean that a task needs running which was scheduled.
adding a print in the `for (uid,generator) in self._async[0]:` loop should be able to get you the name of `uid` - it should really help understand what is causing it.

**Comment by chantra** [comment]
I will try that. This is most likely going to generate tons of logs, but I may be able to gate that behind a file marker, and only log when the file is present.

**Comment by chantra** [comment]
ok, so I have some form of repro but not much actionable yet.

Setup:
exabgp running with `exabgp.conf`
```
template {
  neighbor all {
    router-id 172.18.0.3;
    local-as 65001;
    hold-time 20;
    capability {
      graceful-restart;
    }
  }
}

#process foo {
#  run '/mnt/hgfs/devtools/exabgp/bin/showadjrib.py';
#  encoder text;
#}

neighbor 172.18.0.2 {
 inherit all;
  local-address 172.18.0.3;
  family {
    ipv4 unicast;
    ipv6 unicast;
  }
  peer-as 65000;
}
```

`exabgp.env`:
```
[exabgp.daemon]
drop = false
user = 'root'

[exabgp.tcp]
bind = '0.0.0.0'

[exabgp.log]
level=DEBUG
```


on 172.18.0.2, there is bird running with this config:

```
router id 172.18.0.2;
protocol kernel {
	scan time 60;
	import none;
}
protocol device {
	scan time 60;
}
protocol bgp bgp1{
	import all;
	local as 65000;
	neighbor 172.18.0.3 as 65001;
}
```

steps:
* start bird
* start exabgp to peer with bird.
* send `kill -SIGSTOP ${birdpid}`
* wait for exabgp to `initialising connection to peer-1` while bird is stopped.

CPU goes 100%

**Comment by chantra** [comment]
here some profiling files in case that help. I am still digging through attempting to find what triggers the busyloop (which ends up being repetitive calls to `select.select`

## Kcachegrind versions
[prof-nok.log](https://github.com/Exa-Networks/exabgp/files/1360905/prof-nok.log)
[prof-ok.log](https://github.com/Exa-Networks/exabgp/files/1360906/prof-ok.log)

## pstats versions
[prof-nok.log.prof.log](https://github.com/Exa-Networks/exabgp/files/1360910/prof-nok.log.prof.log)
[prof-ok.log.prof.log](https://github.com/Exa-Networks/exabgp/files/1360909/prof-ok.log.prof.log)

**Comment by chantra** [comment]
ok, even simpler wait to reproduce.... instead of running `bird` on the other end, you can just run:

```
nc -l -p 179
```

basically, TCP connection get established, OPEN get sent and we busy loop waiting to read from the peer.

I tried setting `open.bgp.openwait` to something low like `2`, hoping that we would get the busy spin for 2 sec and then it would stop until the next connection establishes, but it was still taking 100% CPU.

For now, I can narrow this down to https://github.com/Exa-Networks/exabgp/blob/master/lib/exabgp/reactor/peer.py#L323

It is unclear to me if this is the original issue, but this is definitely something that can contribute to it :)

**Comment by chantra** [comment]
@thomas-mangin so... This is not great, but something like https://gist.github.com/chantra/b774ae2e6661a6cc52e1b48bd4b33f73 may mitigate the issue for now.
I checked the 3.4 code and while for the most part some of the core logic is similar, I was not able to reproduce the issue on it.

**Comment by thomas-mangin** [comment]
This will indeed mitigate the issue .. I need to perform the proper fix. Hopefully I will have time this week-end.

**Comment by chantra** [comment]
Thanks @thomas-mangin ... I will keep an eye on how this workaround behaves but for now, I think this is a much better position than pegging a core.

Let me know if there is anything I can help with. The event loop/state machine is non trivial and I would need more time to wrap my head around how things fit together.

Shouldn't this be deal at the loop.py level directly? as it is, it looks like thee is basically 2 event loop running in parallel.

**Comment by chantra** [comment]
So, the patch in #730 will get us on par with the old behaviour of 3.4-stable, e.g yield ACTION.LATER when reading OPEN.

Now, this unveils another issue when the notification get sent:
https://github.com/Exa-Networks/exabgp/blob/master/lib/exabgp/reactor/peer.py#L282

again, ExaBGP process will pin 1 core. 3.4-stable does not have the issue.

To repro the latest issue, assuming #730 is applied:

1. on peer `nc -l 179`
1. on exabgp, set `exabgp.bgp.openwait` to a small value to not wait for too long
1. run exabgp, wait for notification to be sent, CPU is at 100%


```
03:12:58 | 1742   | reactor       | loaded new configuration successfully
03:12:58 | 1742   | process       | forked process api-internal-cli-95bdc012
03:12:58 | 1742   | reactor       | initialising connection to peer-1
03:12:58 | 1742   | outgoing-1    | attempting connection to 172.18.0.2:179
03:12:58 | 1742   | outgoing-1    | sending TCP payload (  67) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 0043 0104 FDE9 0014 AC12 0003 2602 0C40 0A80 1400 0101 8000 0201 8002 0601 0400 0100 0102 0601 0400 0200 0102 0641 0400 00FD E9
03:12:58 | 1742   | outgoing-1    | >> OPEN version=4 asn=65001 hold_time=20 router_id=172.18.0.3 capabilities=[Multiprotocol(ipv4 unicast,ipv6 unicast), Graceful Restart Flags 0x8 Time 20 ipv4/unicast=0x80 ipv6/unicast=0x80, ASN4(65001)]
03:12:58 | 1742   | ka-outgoing-1 | receive-timer 4 second(s) left
03:12:59 | 1742   | ka-outgoing-1 | receive-timer 3 second(s) left
03:13:00 | 1742   | ka-outgoing-1 | receive-timer 2 second(s) left
03:13:01 | 1742   | ka-outgoing-1 | receive-timer 1 second(s) left
03:13:02 | 1742   | ka-outgoing-1 | receive-timer 0 second(s) left
03:13:02 | 1742   | outgoing-1    | sending TCP payload (  77) FFFF FFFF FFFF FFFF FFFF FFFF FFFF FFFF 004D 0301 0177 6169 7465 6420 666F 7220 6F70 656E 2074 6F6F 206C 6F6E 672C 2077 6520 646F 206E 6F74 206C 696B 6520 7374 7563 6B20 696E 2061 6374 6976 65
03:13:02 | 1742   | outgoing-1    | >> NOTIFICATION (1,1,"waited for open too long, we do not like stuck in active")
03:13:02 | 1742   | outgoing-1    | peer reset, message [notification sent (1,1)] error[Message header error / Connection Not Synchronized / waited for open too long, we do not like stuck in active]
03:13:02 | 1742   | outgoing-1    | outgoing-1 172.18.0.3-172.18.0.2, closing connection
03:13:02 | 1742   | reactor       | initialising connection to peer-1
03:13:02 | 1742   | outgoing-2    | attempting connection to 172.18.0.2:179
03:13:02 | 1742   | outgoing-2    | connection to 172.18.0.2 closed
03:13:02 | 1742   | outgoing-2    | outgoing-2 172.18.0.3-172.18.0.2, closing connection
03:13:03 | 1742   | reactor       | initialising connection to peer-1
03:13:03 | 1742   | outgoing-3    | attempting connection to 172.18.0.2:179
```

**Comment by thomas-mangin** [comment]
Hello @chantra - yes the patch in #730 does look better and would be acceptable with a comment explaining why it was reverted.. still trying to see if there is not a better way -

**Comment by thomas-mangin** [comment]
> Now, this unveils another issue when the notification get sent:
> https://github.com/Exa-Networks/exabgp/blob/master/lib/exabgp/reactor/peer.py#L282

Could you elaborate please ?

**Comment by chantra** [comment]
So this is not necessarily the notification itself, and the connection seems to be actually closed:
```
03:13:02 | 1742   | outgoing-1    | outgoing-1 172.18.0.3-172.18.0.2, closing connection
```

but as this kicks in, the CPU goes 100% again.

**Comment by chantra** [comment]
The issue got auto-closed as I had tagged #730 as fixing this one.... well, it fixes 1 part of it. Now, it still goes 100% CPU once we pass the openwait time as explained in https://github.com/Exa-Networks/exabgp/issues/723#issuecomment-334980224 .

Reopening so we can keep on tracking the issue here.

**Comment by chantra** [comment]
Tried it out bc39722e31e0b46a26e086708d849fec16537737 quickly and it still repro :s

**Comment by chantra** [comment]
Thanks for checking it out. From my end, the original dirty hack seems good enough for now, but this issue will be a blocker for 4.0 IMHO

**Comment by chantra** [comment]
This issue is still happening and is most likely happening somewhere deep down in the event loop/state machine of bgp connections.

## Setup
- 2 hosts
  - (peer) with ip 172.18.0.2/16
  - (exabgp) with ip 172.18.0.3/16
- configs on exabgp host:
  - [/etc/exabgp/exabgp.env](https://gist.github.com/chantra/546bffd0e8c50604082c4547a3e652d2)
  - [/etc/exabgp/exabgp.conf](https://gist.github.com/chantra/b2760ebcd4c6a21f5ca13facfb084c43)
  - [Dockerfile](https://gist.github.com/chantra/c22332788695f9f9ea309d79aa0389b3) to generate the exabgp host:

## Repro

On the `peer` host run:
```
nc -l -p 179
```

On the `exabgp` host run (command + output) https://gist.github.com/chantra/e8bb09039579340b94036902e8a8a8d1 

## Observations

Right after the peer reset is sent, exabgp will busy spin.

**Comment by thomas-mangin** [comment]
could https://github.com/Exa-Networks/exabgp/commit/da0be543a46bb4382fa1ebdc99f8e7f92dd4c590 have helped ?

**Comment by chantra** [comment]
@thomas-mangin I gave it a try but the issue is still there.

**Comment by thomas-mangin** [comment]
thanks - I will continue my bug hunting ..

**Comment by chantra** [comment]
yes, this is non-trivial I suppose. There may be a slight change between 3.4 and 4.0 that triggers this.

**Comment by thomas-mangin** [comment]
This patch make me hopeful .. https://github.com/Exa-Networks/exabgp/commit/5a4d95b5829fde6e32f17213885e526d0ecec378

**Comment by pubyun** [comment]
upgrade to git version 47b9b5f, the bug disapeared.

don't know which patch fixed it.

any one confirm it?

**Comment by chantra** [comment]
Interesting. I will try to see if this is fixed for me too.
@pubyun did you get a chance to hit bisect to see what commit may have
fixed this?

**Comment by chantra** [comment]
unfortunately, still reproduces for me on ce5a658b2ba238e6ae26b0a21fa65fb3790b0647 using nc on the peer as described in https://github.com/Exa-Networks/exabgp/issues/723#issuecomment-334980224

**Comment by chantra** [comment]
as I was tryinmg to do some bissection to find when the issue was introduced.... I found out that 331a9b49ace52d05007c8f8f7359b00fde9d1408 (my patch which fixed part of this busy spin) actually introduce the busyspin which is reproduceable with `nc -l -p 179` :s

**Comment by thomas-mangin** [comment]
That’s very useful information

**Comment by thomas-mangin** [comment]
This patch seems to have fixed the issue AFAICS.

**Comment by thomas-mangin** [comment]
@chantra - trying with `nc` does not use 100% CPU anymore when you are happy I will close (no rush) - I just need to add an entry in the CHANGELOG before the release of 4.0.7

**Comment by thomas-mangin** [comment]
I think fixed it one day ago :p

**Comment by thomas-mangin** [comment]
I will check the OPEN case later on today.

**Comment by thomas-mangin** [comment]
A bit more convoluted than I expected ... This time it should be all good.

**Comment by chantra** [comment]
Cool stuff. I will check it back after the long weekend!

**Comment by chantra** [comment]
@thomas-mangin looks like it is working for me too! great job and thanks!
